### PR TITLE
ENH: Connected to  AWS 

### PR DESCRIPTION
**Pull request by KshitizRana**
Created
- authorisation to AWS S3
- Method to upload files to S3
- Method to download from S3
- Added ETL processes
- Aggregated Data

### CODE REVIEW

File: src/database_connect.py [hunk 1]
```diff
@@ -42,4 +43,13 @@
         cr.execute(sql, tuple(row))
         db.commit() 
         if cr.rowcount == 1:
-            total += 1
\ No newline at end of file
+            total += 1
+       
+#Connection to the AWS S3     
+s3 = aws_auth('ap-south-1')
```
Comment (mitulds) [inline src/database_connect.py:49]: remove this

File: src/utils.py [hunk 2]
```diff
@@ -157,4 +164,50 @@ def convert_dtypes(df):
...
+    """Method to download file from AWS S3
+
+    Parameters
+    ----------
+    s3 : 
+    bucket : Name of bucket to download from.
+    key_name : The name of the key to download from.
+    file_name : The name to the file to download
+    """
+    s3.download_file(bucket,key_name,file_name)
```
Comment (mitulds) [inline src/utils.py:196]: replace `download_file` with `get_object`

File: src/utils.py [hunk 3]
```diff
@@ -157,4 +164,50 @@ def convert_dtypes(df):
...
+
+    Parameters
+    ----------
+    s3 : Connection to bucket.
+    file : Filepath for file to be uploaded to bucket.
+    bucket_name : Bucket name where file to be uploaded.
+    object_name : Key to the file to identify in bucket.
+    """
+    # Bucket = 'dtop-project'
+    s3.Client.upload_file(file, bucket_name, object_name)
```
Comment (mitulds) [inline src/utils.py:181]: use `put_object` instead with args 
```
Body= csv_data,
            Bucket = bucket_name,
            Key=f'{filename}.csv'
```

File: src/database_connect.py [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import argparse
 
-from utils import convert_dtypes, data, database, dbconnect, table
+from utils import convert_dtypes, data, database, dbconnect, table, aws_auth, upload_to_s3, download_from_s3
```
Comment (mitulds) [inline src/database_connect.py:3]: you don't need to import all the functions, only import the ones that you are using inside the script

File: src/etl.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        DataFrame: resultant dataframe with hourly timestamps.
+    """
+    dummy = df.copy()
+    dummy[column] = pd.to_datetime(dummy[column], format='%Y-%m-%d %H:%M:%S')  # String to datetime datatype conversion
+    dummy[column] = dummy[column].dt.floor('h')  # Truncate timestamps to beginning of hour
+    return dummy
+    
+
+# Aggregate data
+sales_df = convert_timestamp_to_hourly(pd.read_csv('data/sales.csv'),'timestamp')
```
Comment (mitulds) [inline src/etl.py:23]: instead of using `read_csv`, use `data` function from your utils

File: src/utils.py [hunk 6]
```diff
@@ -157,4 +163,39 @@ def convert_dtypes(df):
...
+    """
+    s3 = aws_auth()
+    csv_data = df.to_csv(index = False)
+    s3.put_object(Body = csv_data,
+                  Bucket = bucket_name,
+                  Key = f'{file_name}.csv')
+    logger.info(f'Successfully Uploaded data in {bucket_name} as {file_name}')
+    
+
+def download_from_s3(bucket,key_name,file_name):
```
Comment (mitulds) [inline src/utils.py:187]: this function is incomplete.
When you do get_object, it returns a json object. You need to store that json object into a variable and extract data from it.

```
response = s3.get_object(bucket,key_name,file_name)
df = pd.read_csv(response['Body'])
return df
```

**Comment by mitulds** [comment]
@KshitizRana left some feedback to work on

File: src/etl.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    dummy[column] = dummy[column].dt.floor('h')  # Truncate timestamps to beginning of hour
+    return dummy
+    
+
+# Aggregate data
+sales_df = convert_timestamp_to_hourly(data('data/sales.csv'),'timestamp')
+salesagg_df = sales_df.groupby(by=['timestamp','product_id']).agg({'quantity' : 'sum'}).reset_index()
+salesagg_df.to_csv('data/sales_agg.csv',index=False)
+
+stocklevel_df = convert_timestamp_to_hourly(pd.read_csv('data/sensor_stock_levels.csv'),'timestamp')
```
Comment (mitul23) [inline src/etl.py:28]: You haven't replace read_csv everywhere

File: src/utils.py [hunk 8]
```diff
@@ -157,4 +163,42 @@ def convert_dtypes(df):
...
+    file_name : The name to the file to download
+    """
+    s3 = aws_auth()
+    response = s3.get_object(bucket,key_name,file_name)
+    df = pd.read_csv(response['Body'])
+    logger.info(f'Successfully Downloaded data from {bucket} as {file_name}')   
+    
+    return df
+    
+
```
Comment (mitul23) [inline src/utils.py:204]: Remove empty lines from the end of the file

**Review by mitul23 — APPROVED** [review]
Few minor changes, otherwise good to merge!

**Review by mitulds — APPROVED** [review]

File: src/etl.py (hunk 7 above)
Comment (mitulds) [reply src/etl.py:28]: best to do relevant changes in the same PR

**Pull request merged**